### PR TITLE
fix #1985 Add FAQ section about logging with MDC

### LINKED ISSUE #1985: Guidance on logging with MDC
The Reactor documentation has [Adding Context To a Reactive Sequence](https://projectreactor.io/docs/core/release/reference/#context) that covers the topic of `ThreadLocal` alternatives but there is no mention of how to log with MDC context.

@simonbasle has a [blog post](https://simonbasle.github.io/2018/02/contextual-logging-with-reactor-context-and-mdc/) but I think it must not be easy to find because various suggestions keep showing up that typically tend to not work in all cases or have significant trade-offs.

There should be some guidance that is easier to find. Possibly even a short section in the Reactor docs with a link to the blog post. Whatever it ends up being, we can further link to it from the Spring Framework docs.

A more ambitious option would be to provide logging framework integrations but that could end up as a separate issue. In the mean time some minimal documentation would really help.

### PR DESCRIPTION
**Pull request by simonbasle**

### CODE REVIEW

**Comment by simonbasle** [comment]
cc also @Buzzardo if you have time to review 🙇

File: docs/asciidoc/faq.adoc [hunk 1]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
          .subscribe();
 ----
 
+== What is a good pattern for contextual logging? (MDC)
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Headings should be in title case, so this should be "What Is a Good Pattern for Contextual Logging? (MDC)".

File: docs/asciidoc/faq.adoc [hunk 2]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+As we saw in section <<context>>, `ThreadLocal` doesn't make much sense in Reactive Streams, where several streams could execute steps on a common thread.
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Change "As we saw" to "As described" (because the reader may not have seen that yet). Change "doesn't" to "does not" (formal documentation does not use contractions).  Change "execute" to "run" (less violent connotation).

File: docs/asciidoc/faq.adoc [hunk 3]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+Instead of using `ThreadLocal`, Reactor offers the `Context`, a map-like object that is tied to a `Subscription` and not a `Thread`.
+
+One recurring use of `ThreadLocal` in java is for logging frameworks to store variables that will be used in their logging pattern, generally by way of a `Map` called the MDC ("Mapped Diagnostic Context").
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Capitalize "java". Change "will be" to "are" (keeping the tense consistent throughout).

File: docs/asciidoc/faq.adoc [hunk 4]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+
+Unfortunately, this pattern is hard to opt out of and hard to apply transparently from a reactive library perspective.
+This entry of the FAQ offers a possible intermediate solution when one wants to perform log statements in relation to events in the stream (`onNext`, `onError`, `onComplete`).
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Add "and" before "`onComplete`".

File: docs/asciidoc/faq.adoc [hunk 5]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+
+The most straightforward way to perform such logging and get contextual information from the `Context` to the MDC is to wrap logging statements in a `doOnEach` operator with a little bit of boilerplate code.
+This boilerplate depends on both the logging framework/abstraction of your choice and the information you want to put in the MDC, so it has to be in your codebase.
+
+The following is an example of such a helper function around a single MDC variable and focused on logging `onNext` events, using Java 9 enhanced `Optional` API:
+[source,java]
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Above the "[source,java]" line, add another line with four equal signs ("===="). We want to contain the code callouts within a bounded box.

File: docs/asciidoc/faq.adoc [hunk 6]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+		() -> logStatement.accept(signal.get())); <5>
+	};
+}
+----
+<1> `doOnEach` signals include onComplete and onError. In this example we're only interested in logging `onNext`
+<2> We will extract one interesting value from the Reactor `Context` (see the <<context.api>> section)
+<3> We use the `MDCCloseable` from SLF4J 2 in this example, allowing try-with-resource syntax for automatic cleanup of the MDC after the log statement is executed
+<4> Proper log statement is provided by the caller as a `Consumer<T>` (consumer of the onNext value)
+<5> In case the expected key wasn't set in the `Context` we use the alternative path where nothing is put in the MDC
+
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Add four equal signs ("====") here, to close the box that encloses both the code and the callouts.

File: docs/asciidoc/faq.adoc [hunk 7]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+		toPutInMdc.ifPresentOrElse(tpim -> {
+			try (MDC.MDCCloseable cMdc = MDC.putCloseable("MDC_KEY", tpim)) { <3>
+				logStatement.accept(signal.get()); <4>
+			}
+		},
+		() -> logStatement.accept(signal.get())); <5>
+	};
+}
+----
+<1> `doOnEach` signals include onComplete and onError. In this example we're only interested in logging `onNext`
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Add backticks around "onComplete" and "onError" (so "`onComplete` and `onError`"). Change "we're" to "we are" (no contractions). Add a period at the end of the sentence.

File: docs/asciidoc/faq.adoc [hunk 8]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+<2> We will extract one interesting value from the Reactor `Context` (see the <<context.api>> section)
+<3> We use the `MDCCloseable` from SLF4J 2 in this example, allowing try-with-resource syntax for automatic cleanup of the MDC after the log statement is executed
+<4> Proper log statement is provided by the caller as a `Consumer<T>` (consumer of the onNext value)
+<5> In case the expected key wasn't set in the `Context` we use the alternative path where nothing is put in the MDC
+
+Using this boilerplate code ensures that we are good citizen with the MDC: we set a key right before we execute a logging statement and remove it immediately after.
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: "citizen" should be "citizens" (plural), to match "we".

File: docs/asciidoc/faq.adoc [hunk 9]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+There is no risk of polluting the MDC for subsequent logging statements.
+
+Of course, this is a suggestion. You might be interested in extracting multiple values from the `Context`, or in also logging things in case of `onError`.
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Remove the comma between "`Context`" and "or". You don't need commas when the list has two items. Remove "also" (being in a list makes it unnecessary, and I had to read the sentence twice to figure out what was going on).

File: docs/asciidoc/faq.adoc [hunk 10]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+You might want to create additional helper methods for these cases, or craft a single method that makes use of additional lambdas to cover more ground.
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Remove the comma (not allowed in a two-item list).

File: docs/asciidoc/faq.adoc [hunk 11]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+
+In any case, the usage of the above helper method could look like the following reactive web controller:
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Change "above" to "preceding" (in PDF or some other paged media, the helper method may not be "above").

File: docs/asciidoc/faq.adoc [hunk 12]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+
+[source,java]
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Add the "====" fences around this code block, partly because it has callouts and partly to be consistent.

File: docs/asciidoc/faq.adoc [hunk 13]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+public Flux<Restaurant> byPrice(@RequestParam Double maxPrice, @RequestHeader(required = false, name = "X-UserId") String userId) {
+	String apiId = userId == null ? "" : userId; <1>
+
+	return restaurantService.byPrice(maxPrice))
+			   .doOnEach(logOnNext(r -> LOG.debug("found restaurant {} for ${}", <2>
+					r.getName(), r.getPricePerPerson())))
+			   .subscriberContext(Context.of("CONTEXT_KEY", apiId)); <3>
+}
+----
+<1> We will need to get the contextual information from the request header to put it in the `Context`
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Remove "will" (keeping tense consistent).

File: docs/asciidoc/faq.adoc [hunk 14]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+<2> Here we apply our helper method to the `Flux` using `doOnEach`. Remember operators see `Context` values defined below them.
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Add either a comma or "by" before "using" (I would add a comma). Add either a colon or "that" after "Remember". I would add a colon.

File: docs/asciidoc/faq.adoc [hunk 15]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+<3> We write the value from the header to the `Context` using the chosen key `CONTEXT_KEY`.
+
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Add the "====" fence here, to match the same line before the sample.

File: docs/asciidoc/faq.adoc [hunk 16]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+In this configuration, the `restaurantService` can emit its data on a shared thread yet the logs will still reference the correct `X-UserId` for each request.
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Add a comma before "yet".

File: docs/asciidoc/faq.adoc [hunk 17]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+
+For completeness, let's also see what an error-logging helper could look like:
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Change "let's" to "we can". ("let's" is a problem because of the prohibition on contractions, and "let us" always sounds weird to me, so I generally recast these sentences.)

File: docs/asciidoc/faq.adoc [hunk 18]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Add the "====" fences around this code block, even though there are no callouts (to be consistent with the other code blocks).

File: docs/asciidoc/faq.adoc [hunk 19]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+		toPutInMdc.ifPresentOrElse(tpim -> {
+			try (MDC.MDCCloseable cMdc = MDC.putCloseable("MDC_KEY", tpim)) {
+				errorLogStatement.accept(signal.getThrowable());
+			}
+		},
+		() -> errorLogStatement.accept(signal.getThrowable()));
+	};
+}
+----
+
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Add the closing fence: "====".

File: docs/asciidoc/faq.adoc [hunk 20]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+Nothing much has changed except for the fact that we check the `Signal` is effectively an `onError`, and we provide said error (a `Throwable`) to the log statement lambda.
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Add a comma between "changed" and "except". Add that between "check" and "the".

File: docs/asciidoc/faq.adoc [hunk 21]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+
+Applying this helper in the controller is very similar to what we've done before:
+
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: "====" around the block and the callout.

File: docs/asciidoc/faq.adoc [hunk 22]
```diff
@@ -206,4 +206,102 @@ processor.publishOn(scheduler1)
...
+	String apiId = userId == null ? "" : userId;
+
+	return restaurantService.byPrice(maxPrice))
+			   .doOnEach(logOnNext(v -> LOG.info("found restaurant {}", v))
+			   .doOnEach(logOnError(e -> LOG.error("error when searching restaurants", e)) <1>
+			   .subscriberContext(Context.of("CONTEXT_KEY", apiId));
+}
+----
+<1> In case the `restaurantService` emits an error, it will be logged with MDC context here
+
```
Comment (Buzzardo) [inline docs/asciidoc/faq.adoc]: Add the closing "====".

File: docs/asciidoc/faq.adoc (hunk 4 above)
Comment (simonbasle) [reply docs/asciidoc/faq.adoc]: I'm assuming keeping the oxford comma for consistency, even though here it could be considered redundant since the list is not ambiguous?

**Comment by simonbasle** [comment]
@Buzzardo thanks for the review, I have pushed changes accordingly

File: docs/asciidoc/faq.adoc [hunk 23]
```diff
@@ -205,5 +226,113 @@ processor.publishOn(scheduler1)
          .doOnNext(i -> processNext(i))
          .subscribe();
 ----
+====
+
+== What Is a Good Pattern for Contextual Logging? (MDC)
+As described in section <<context>>, `ThreadLocal` does not make much sense in Reactive Streams, where several streams could run steps on a common thread.
+Instead of using `ThreadLocal`, Reactor offers the `Context`, a map-like object that is tied to a `Subscription` and not a `Thread`.
+
```
Comment (rstoyanchev) [inline docs/asciidoc/faq.adoc]: Perhaps it would be more helpful to describe this in the context of a functional, declarative API rather than Reactive Streams? Those are meant to be thread agnostic, and each operator/stage could execute in a different thread. Then it's also easier to talk about passing a context through the composition chain as a more idiomatic solution in a functional API, and this is why Reactor has a built-in `Context` that propagates throughout the execution chain as long as Flux or Mono is used as the return value, etc.

So that takes you pretty far, but within a reactive component you might still end up calling something that doesn't have reactive types in its signature and for MDC logging context might become a necessary solution.

In other words perhaps worth phrasing it in such a way as to emphasize that MDC shouldn't be an automatic assumption, something that is all too easy to assume.

File: docs/asciidoc/faq.adoc [hunk 24]
```diff
@@ -205,5 +226,113 @@ processor.publishOn(scheduler1)
...
+One recurring use of `ThreadLocal` in Java is for logging frameworks to store variables that are used in their logging pattern, generally by way of a `Map` called the MDC ("Mapped Diagnostic Context").
+
+Unfortunately, this pattern is hard to opt out of and hard to apply transparently from a reactive library perspective.
+This entry of the FAQ offers a possible intermediate solution when one wants to perform log statements in relation to events in the stream (`onNext`, `onError`, and `onComplete`).
+
+The most straightforward way to perform such logging and get contextual information from the `Context` to the MDC is to wrap logging statements in a `doOnEach` operator with a little bit of boilerplate code.
+This boilerplate depends on both the logging framework/abstraction of your choice and the information you want to put in the MDC, so it has to be in your codebase.
+
```
Comment (rstoyanchev) [inline docs/asciidoc/faq.adoc]: One thing that might puzzle someone, who has not used the context, is how that context is inserted in the first place. Perhaps worth pointing out that the context key needs to be inserted early in the execution chain, possibly through a `WebFilter` (e.g. in a WebFlux application) so that it is available throughout. Or if there is a convenient place to link to for learning more on how to insert context.

File: docs/asciidoc/faq.adoc [hunk 25]
```diff
@@ -205,5 +226,113 @@ processor.publishOn(scheduler1)
...
+====
+[source,java]
+----
+public static <T> Consumer<Signal<T>> logOnNext(Consumer<T> logStatement) {
+	return signal -> {
+		if (!signal.isOnNext()) return; <1>
+		Optional<String> toPutInMdc = signal.getContext().getOrEmpty("CONTEXT_KEY"); <2>
+
+		toPutInMdc.ifPresentOrElse(tpim -> {
```
Comment (rstoyanchev) [inline docs/asciidoc/faq.adoc]: Typo? Should this be just `ifPresent`?

File: docs/asciidoc/faq.adoc [hunk 26]
```diff
@@ -205,5 +226,113 @@ processor.publishOn(scheduler1)
...
+
+====
+[source,java]
+----
+public static Consumer<Signal<?>> logOnError(Consumer<Throwable> errorLogStatement) {
+	return signal -> {
+		if (!signal.isOnError()) return;
+		Optional<String> toPutInMdc = signal.getContext().getOrEmpty("CONTEXT_KEY");
+
+		toPutInMdc.ifPresentOrElse(tpim -> {
```
Comment (rstoyanchev) [inline docs/asciidoc/faq.adoc]: Just `ifPresent`?

File: docs/asciidoc/faq.adoc (hunk 25 above)
Comment (simonbasle) [reply docs/asciidoc/faq.adoc]: nope, using the "enhanced Optional API" from Java 9 to cover the case where there's no interesting value in the `Context` (second lambda, no need to feed the MDC in that case)

File: docs/asciidoc/faq.adoc (hunk 26 above)
Comment (simonbasle) [reply docs/asciidoc/faq.adoc]: (see above)

File: docs/asciidoc/faq.adoc (hunk 24 above)
Comment (simonbasle) [reply docs/asciidoc/faq.adoc]: added a small hint at that with a more explicit recommendation to read the `<<Context>>` section

File: docs/asciidoc/faq.adoc (hunk 23 above)
Comment (simonbasle) [reply docs/asciidoc/faq.adoc]: reworked the introduction to address this :+1:

**Comment by simonbasle** [comment]
I think all has been resolved in this PR. Please re-open the issue if that is not satisfactory @rstoyanchev
